### PR TITLE
PWN-7352 - Fix history scrolling behaviour, optimize main screen requ…

### PR DESCRIPTION
**Pull request by Davran331122**
…ests

## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7352

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/history/ui/token/adapter/HistoryAdapter.kt [hunk 1]
```diff
@@ -67,23 +49,27 @@ class HistoryAdapter(
...
+        return when (val item = currentList.getOrNull(position)) {
             is TransactionItem -> item.transactionId.hashCode().toLong()
             is DateItem -> item.date.hashCode().toLong()
             else -> RecyclerView.NO_ID
         }
     }
 
+    override fun setHasStableIds(hasStableIds: Boolean) {
+        super.setHasStableIds(true)
+    }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/ui/token/adapter/HistoryAdapter.kt:71]: you can set it from fragment

File: app/src/main/java/org/p2p/wallet/common/di/AppScope.kt [hunk 2]
```diff
@@ -6,5 +6,5 @@ import kotlinx.coroutines.SupervisorJob
 import kotlin.coroutines.CoroutineContext
 
 class AppScope : CoroutineScope {
-    override val coroutineContext: CoroutineContext = SupervisorJob() + Dispatchers.Main
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/common/di/AppScope.kt:9]: blet

File: app/src/main/java/org/p2p/wallet/common/di/AppScope.kt [hunk 3]
```diff
@@ -6,5 +6,5 @@ import kotlinx.coroutines.SupervisorJob
...
+    override val coroutineContext: CoroutineContext = SupervisorJob() + Dispatchers.IO
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/common/di/AppScope.kt:9]: Not sure about this

**Review by 1sakov — APPROVED** [review]

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/common/di/AppScope.kt (hunk 3 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/common/di/AppScope.kt:9]: Me too, but it's a little bit confusing to use main thread for long term tasks,
we can discuss about it.

File: app/src/main/java/org/p2p/wallet/common/feature_toggles/remote_config/AppFirebaseRemoteConfig.kt [hunk 4]
```diff
@@ -20,28 +24,30 @@ class AppFirebaseRemoteConfig : RemoteConfigValuesProvider {
...
-
-        // maybe add loading screen in the start for remoteConfig to finish the job?
-        remoteConfig.fetchAndActivate()
-            .addOnSuccessListener {
-                isFetchFailed = false
-                Timber.d("Remote config fetched and activated")
-                Timber.i("Remote config fetched toggles: ${allFeatureTogglesRaw()}")
-
-                onConfigLoaded.invoke(allFeatureTogglesRaw())
+        appScope.launch {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/common/feature_toggles/remote_config/AppFirebaseRemoteConfig.kt:27]: it's better to launch this task in the place where it called

1) AppCreatedAction already has AppScope
2) Fetching remote config is async task by default, launch is redundant

File: app/src/main/java/org/p2p/wallet/history/HistoryModule.kt [hunk 5]
```diff
@@ -83,7 +82,7 @@ object HistoryModule : InjectionModule {
                     hiddenSellTransactionsStorageContract = get()
                 )
             )
-            HistoryRepository(remotes)
+            HistoryRepository(remotes, get())
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/HistoryModule.kt:85]: named args

File: app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt [hunk 6]
```diff
@@ -1,20 +1,25 @@
...
-    private val historyServiceRepository: HistoryRemoteRepository
+    private val historyServiceRepository: HistoryRemoteRepository,
+    private val coroutineDispatchers: CoroutineDispatchers
 ) {
 
-    suspend fun loadHistory(limit: Int, mintAddress: String?): HistoryPagingResult {
-        return historyServiceRepository.loadHistory(limit, mintAddress)
-    }
+    suspend fun loadHistory(limit: Int, mintAddress: String?): HistoryPagingResult =
+        withContext(coroutineDispatchers.io) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt:15]: withContext is used ONLY in repositories

File: app/src/main/java/org/p2p/wallet/history/model/rpc/RpcHistoryTransaction.kt [hunk 7]
```diff
@@ -72,6 +72,10 @@ sealed class RpcHistoryTransaction(
             return "${amount.total.abs().scaleLong().toPlainString()} $tokenSymbol"
         }
 
+        override fun getSymbol(isNegativeOperation: Boolean): String {
+            return if (isNegativeOperation) "" else "+"
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/model/rpc/RpcHistoryTransaction.kt:76]: "" -> emptyString()

File: app/src/main/java/org/p2p/wallet/history/repository/remote/HistoryRepository.kt [hunk 8]
```diff
@@ -37,26 +40,27 @@ class HistoryRepository(
...
+                    }
                 }
             }
+            val errorMessage = errorMessageBuilder.toString()
+            if (errorMessage.isNotEmpty()) {
+                return@withContext HistoryPagingResult.Error(Throwable(errorMessage))
+            }
+            return@withContext newTransactions.sortedByDescending { transaction ->
+                transaction.date.dateMilli()
+            }.let(HistoryPagingResult::Success)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/HistoryRepository.kt:64]:
```suggestion
            if (errorMessage.isNotEmpty()) {
                HistoryPagingResult.Error(Throwable(errorMessage))
            } else {
                newTransactions
                    .sortedByDescending { transaction -> transaction.date.dateMilli() }
                    .let(HistoryPagingResult::Success)
           }
```

File: app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRepository.kt [hunk 9]
```diff
@@ -23,33 +25,36 @@ class RpcHistoryRepository(
...
-            val response = fetchHistoryTransactions(limit, mintAddress)
-            val domainItems = response.map { converter.toDomain(it) }
-            HistoryPagingResult.Success(domainItems)
-        } catch (e: Throwable) {
-            HistoryPagingResult.Error(e)
+    override suspend fun loadHistory(limit: Int, mintAddress: String?): HistoryPagingResult =
+        withContext(coroutineDispatchers.io) {
+            allTransactions.clear()
+            tokenPagingState.clear()
+            return@withContext try {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRepository.kt:39]: no need for return here

File: app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRepository.kt [hunk 10]
```diff
@@ -23,33 +25,36 @@ class RpcHistoryRepository(
...
-    override suspend fun loadNextPage(limit: Int, mintAddress: String?): HistoryPagingResult {
-        return try {
-            val response = fetchHistoryTransactions(limit, mintAddress)
-            val domainItems = response.map { converter.toDomain(it) }
-            HistoryPagingResult.Success(domainItems)
-        } catch (e: Throwable) {
-            HistoryPagingResult.Error(e)
+    override suspend fun loadNextPage(limit: Int, mintAddress: String?): HistoryPagingResult =
+        withContext(coroutineDispatchers.io) {
+            return@withContext try {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRepository.kt:50]: no need for withContext

File: app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRepository.kt [hunk 11]
```diff
@@ -99,7 +104,7 @@ class RpcHistoryRepository(
         val localTransactions = findTransactionsByTokenAddress(
             token = tokenAddress
         )
-        return try {
+        return@withContext try {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRepository.kt:107]: no need for return

File: app/src/main/java/org/p2p/wallet/history/signature/HistoryServiceSignatureFieldGenerator.kt [hunk 12]
```diff
@@ -1,23 +1,26 @@
...
 import org.p2p.wallet.utils.toBase58Instance
 import java.util.Optional
 
 class HistoryServiceSignatureFieldGenerator(
-    private val pushServiceSignatureFieldGenerator: PushServiceSignatureFieldGenerator
+    private val pushServiceSignatureFieldGenerator: PushServiceSignatureFieldGenerator,
+    private val coroutineDispatchers: CoroutineDispatchers
 ) {
-    fun generateSignature(
+    suspend fun generateSignature(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/signature/HistoryServiceSignatureFieldGenerator.kt:16]: generating signature is made in mappers, mappers are located and called in repositories that are already have withContext on site. 
there are a lot of withContext here and there, code became inconsistent

File: app/src/main/java/org/p2p/wallet/history/signature/HistoryServiceSignatureFieldGenerator.kt [hunk 13]
```diff
@@ -1,23 +1,26 @@
...
         pubKey: String,
         privateKey: ByteArray,
         offset: Long,
         limit: Long,
         mint: Optional<String>
-    ): Base58String {
-        return pushServiceSignatureFieldGenerator.generateSignatureField(
+    ): Base58String = withContext(coroutineDispatchers.io) {
+        return@withContext pushServiceSignatureFieldGenerator.generateSignatureField(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/signature/HistoryServiceSignatureFieldGenerator.kt:23]: no need fo return

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt [hunk 14]
```diff
@@ -26,67 +26,68 @@ class HistoryListView @JvmOverloads constructor(
...
 
     private val binding = inflateViewBinding<LayoutHistoryListBinding>()
 
-    private var tokenForHistory: Token.Active? = null
+    private var tokenMintAddress: String? = null
 
     private val glideManager: GlideManager by inject()
 
-    private lateinit var historyAdapter: HistoryAdapter
+    private val historyAdapter: HistoryAdapter by lazy {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt:37]: by unsafeLazy

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt [hunk 15]
```diff
@@ -26,67 +26,68 @@ class HistoryListView @JvmOverloads constructor(
...
-            glideManager = glideManager,
-            onHistoryItemClicked = presenter::onItemClicked,
-            onRetryClicked = presenter::loadNextHistoryPage,
-        )
         bindView()
-        presenter.attach(this)
     }
 
     private fun bindView() {
+        presenter.attach(this)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt:67]: i think presenter is attached twice

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt [hunk 16]
```diff
@@ -123,14 +124,15 @@ class HistoryListView @JvmOverloads constructor(
         }
     }
 
-    override fun showHistory(history: List<HistoryItem>) {
+    override suspend fun showHistory(history: List<HistoryItem>) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt:127]: show history should not be suspend
use viewLifecycleScope or smth like that

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt [hunk 17]
```diff
@@ -123,14 +124,15 @@ class HistoryListView @JvmOverloads constructor(
...
         with(binding) {
-            historyAdapter.setTransactions(history)
-            historyRecyclerView.invalidateItemDecorations()
-
-            val isHistoryEmpty = historyAdapter.isEmpty()
-            emptyStateLayout.root.isVisible = isHistoryEmpty
-            historyRecyclerView.isVisible = !isHistoryEmpty
+            historyAdapter.submitList(history) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt:129]: named arg for lambda

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt [hunk 18]
```diff
@@ -18,28 +18,22 @@ private const val PAGE_SIZE = 20
...
+        attachToHistoryFlow()
     }
 
-    override fun onResume(owner: LifecycleOwner) {
-        super.onResume(owner)
-        loadHistory()
-    }
-
-    override fun loadNextHistoryPage() {
+    override fun loadNextHistoryPage(mintAddress: String?) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt:30]: which mint address? last token that is loaded? 
lastLoadedTokenMint

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt [hunk 19]
```diff
@@ -48,31 +42,28 @@ class HistoryListViewPresenter(
         }
     }
 
-    override fun loadHistory() {
+    override fun loadHistory(mintAddress: String?) {
         launch {
             try {
                 view?.showPagingState(PagingState.InitialLoading)
-                val result = historyInteractor.loadHistory(PAGE_SIZE, token?.mintAddress)
+                Timber.tag("_____start").d("LoadHistory")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt:49]: revert

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomeElementItemMapper.kt [hunk 20]
```diff
@@ -29,6 +33,6 @@ class HomeElementItemMapper {
             result += hiddenTokens.map { HomeElementItem.Hidden(it, visibilityState) }
         }
 
-        return result
+        return@withContext result
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomeElementItemMapper.kt:36]: no need for that

File: app/src/main/java/org/p2p/wallet/sell/interactor/HistoryItemMapper.kt [hunk 21]
```diff
@@ -33,10 +45,10 @@ class HistoryItemMapper(private val resources: Resources) {
                 }
             }
         }
-        return sellHistoryItems + rpcHistoryItems
+        historyItemFlow.tryEmit(sellHistoryItems + rpcHistoryItems)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/interactor/HistoryItemMapper.kt:48]: why tryEmit? I don't think plus with two lists can throw an exception

**Review by gslevinkov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/sell/interactor/HistoryItemMapper.kt (hunk 21 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/sell/interactor/HistoryItemMapper.kt:48]: tryEmit is async, and emit is suspend function. We should avoid using suspend, and start to using flow for data )

**Pull request merged**